File: LeetCode/1052. Grumpy Bookstore Owner/solution.py
"""
78 / 78 test cases passed.
Runtime: 296 ms
Memory Usage: 17.2 MB
"""
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], X: int) -> int:
        satisfied = sum([cus for i, cus in enumerate(customers) if grumpy[i] == 0])
        slide = sum([customers[i] * grumpy[i] for i in range(X)])
        rest = slide
        for i in range(X, len(grumpy)):
            slide += customers[i] * grumpy[i] - customers[i - X] * grumpy[i - X]
            rest = max(rest, slide)
        return satisfied + rest

File: LeetCode/1052. Grumpy Bookstore Owner/solution.cpp
/**
 * 78 / 78 test cases passed.
 * Runtime: 32 ms
 * Memory Usage: 29.4 MB 
 */
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int X) {
        int sum = 0;
        for (int i = 0; i < customers.size(); ++ i) {
            sum += customers[i] * (1 - grumpy[i]);
        }
        int slide = 0;
        for (int i = 0; i < X; ++ i) {
            slide += customers[i] * grumpy[i];
        }
        int rest = slide;
        for (int i = X; i < grumpy.size(); ++ i) {
            slide += customers[i] * grumpy[i] - customers[i - X] * grumpy[i - X];
            rest = max(rest, slide);
        }
        return sum + rest;
    }
};
